#! /bin/sh /usr/share/dpatch/dpatch-run
## 00_0172-prefer-xrandr-over-xinerama.dpatch by Dirk Mueller
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: only trust libxinerama if its not the emulated information
## DP: coming from xrandr 1.2. xrandr 1.2 is merged fb and libxinerama
## DP: presents then virtual screens in clone mode, which qt (and KDE)
## DP: can't deal with.
## DP: proper fix would be to detect crtcs as virtual screens, but
## DP: given that qt can't deal with dynamically changing number of screens,
## DP: this is for "when I have time".

@DPATCH@
diff -urNad qt4-x11-4.3.0~/src/gui/kernel/qdesktopwidget_x11.cpp qt4-x11-4.3.0/src/gui/kernel/qdesktopwidget_x11.cpp
--- qt4-x11-4.3.0~/src/gui/kernel/qdesktopwidget_x11.cpp	2007-05-25 15:24:19.000000000 +0200
+++ qt4-x11-4.3.0/src/gui/kernel/qdesktopwidget_x11.cpp	2007-06-30 19:52:05.000000000 +0200
@@ -109,11 +109,38 @@
     int unused;
     use_xinerama = (XineramaQueryExtension(X11->display, &unused, &unused) && XineramaIsActive(X11->display));
 
+    // only use xinerama for old Xrandr versions
+#ifndef QT_NO_XRANDR
+    int ncrtc = 0;
+    if (X11->use_xrandr) {
+        int major, minor;
+
+#if RANDR_MAJOR > 1 || RANDR_MINOR > 1
+        XRRQueryVersion(QPaintDevice::x11AppDisplay(), &major, &minor);
+        if (major > 1 || (major == 1 && minor >= 2)) {
+            XRRScreenResources* res;
+            res = XRRGetScreenResources(QPaintDevice::x11AppDisplay(),
+                    QPaintDevice::x11AppRootWindow( 0 ));
+            if (res) {
+                ncrtc = res->ncrtc;
+                XRRFreeScreenResources(res);
+            }
+        }
+#endif
+    }
+#endif
+
     if (use_xinerama) {
         xinerama_screeninfo =
             XineramaQueryScreens(X11->display, &screenCount);
         defaultScreen = 0;
-    } else
+    }
+#ifndef QT_NO_XRANDR
+    if (use_xinerama && screenCount <= ncrtc) {
+        use_xinerama = FALSE;
+    }
+#endif
+    else
 #endif // QT_NO_XINERAMA
     {
         defaultScreen = DefaultScreen(X11->display);
